crates/solidity/testing/solc/src/commands/dissect.rs

```diff
@@ -0,0 +1,172 @@
+use std::path::PathBuf;
+
+use anyhow::Result;
+use clap::Parser;
+use console::{style, Color};
+use infra_utils::paths::PathExtensions;
+use infra_utils::terminal::Terminal;
+use itertools::Itertools;
+use semver::Version;
+use solidity_language::SolidityDefinition;
+
+use crate::utils::{
+    ApiInput, Binary, Error, InputSource, LanguageSelector, Severity, SourceLocation,
+};
+
+#[derive(Debug, Parser)]
+pub struct DissectCommand {
+    /// File path to the Solidity (*.sol) source file to parse
+    file: PathBuf,
+}
+
+impl DissectCommand {
+    pub fn execute(self) -> Result<()> {
+        let language = SolidityDefinition::create();
+        let binaries = Binary::fetch_all(&language)?;
+
+        let mut state = State::new(self.file.to_owned())?;
+
+        for binary in binaries {
+            let input = InputSource {
+                content: state.source.to_owned(),
+            };
+
+            let output = binary.run(&ApiInput {
+                language: LanguageSelector::Solidity,
+                sources: [(self.file.unwrap_string(), input)].into(),
+            });
+
+            let errors = match output {
+                Ok(output) => output.errors.unwrap_or_default(),
+                Err(error) => vec![Error {
+                    message: format!("{error}"),
+                    severity: Severity::Error,
+                    location: None,
+                }],
+            };
+
+            state.apply(binary.version.to_owned(), errors)?;
+        }
+
+        state.print_all()?;
+
+        Ok(())
+    }
+}
+
+struct State {
+    file: PathBuf,
+    source: String,
+
+    errors: Vec<Error>,
+    versions: Vec<Version>,
+}
+
+impl State {
+    fn new(file: PathBuf) -> Result<Self> {
+        let source = file.read_to_string()?;
+
+        Ok(Self {
+            file,
+            source,
+            errors: Vec::new(),
+            versions: Vec::new(),
+        })
+    }
+
+    fn apply(&mut self, version: Version, errors: Vec<Error>) -> Result<()> {
+        if self.versions.is_empty() {
+            assert_eq!(self.errors, vec![], "Impossible state");
+            self.errors = errors;
+            self.versions.push(version);
+        } else if self.errors == errors {
+            self.versions.push(version);
+        } else {
+            self.print_all()?;
+            self.errors = errors;
+            self.versions = vec![version];
+        }
+
+        Ok(())
+    }
```

Comment:
Do you think we could document what's happening and/or why here? I admit that the naming is overly generic, i.e. "state" and "apply" and it takes a second to understand what's going on

Reply:
Thanks. I refactored the module a little bit to make it clearer, and added some comments. Please let me know what you think.